src/xss-network-qsort.hpp

```diff
@@ -0,0 +1,135 @@
+        regs[1] = vtype::reverse(regs[1]);
+        COEX<vtype>(regs[0], regs[1]);
+    }
+    else if constexpr (numVecs > 2) {
+// Reverse upper half
+#pragma GCC unroll 64
+        for (int i = 0; i < numVecs / 2; i++) {
+            reg_t rev = vtype::reverse(regs[numVecs - i - 1]);
+            reg_t maxV = vtype::max(regs[i], rev);
+            reg_t minV = vtype::min(regs[i], rev);
+            regs[numVecs - i - 1] = vtype::reverse(maxV);
+            regs[i] = minV;
+        }
+    }
+
+    // Call cleaner
+    bitonic_clean_n_vec<vtype, numVecs>(regs);
+
+// Now do bitonic_merge
+#pragma GCC unroll 64
+    for (int i = 0; i < numVecs; i++) {
+        regs[i] = vtype::bitonic_merge(regs[i]);
+    }
+}
+
+template <typename vtype,
+          int64_t numVecs,
+          int64_t numPer = 2,
+          typename reg_t = typename vtype::reg_t>
+X86_SIMD_SORT_INLINE void bitonic_fullmerge_n_vec(reg_t *regs)
+{
+    if constexpr (numPer > numVecs)
+        return;
+    else {
+#pragma GCC unroll 64
+        for (int i = 0; i < numVecs / numPer; i++) {
+            bitonic_merge_n_vec<vtype, numPer>(regs + i * numPer);
+        }
+        bitonic_fullmerge_n_vec<vtype, numVecs, numPer * 2>(regs);
+    }
+}
+
+template <typename vtype, int numVecs, typename reg_t = typename vtype::reg_t>
+X86_SIMD_SORT_INLINE void sort_n_vec(typename vtype::type_t *arr, int32_t N)
+{
+    if (numVecs > 1 && N * 2 <= numVecs * vtype::numlanes) {
+        sort_n_vec<vtype, numVecs / 2>(arr, N);
+        return;
+    }
+
+    reg_t vecs[numVecs];
+
+// Unmasked part of the load
+#pragma GCC unroll 64
+    for (int i = 0; i < numVecs / 2; i++) {
+        vecs[i] = vtype::loadu(arr + i * vtype::numlanes);
+    }
+// Masked part of the load
+#pragma GCC unroll 64
+    for (int i = numVecs / 2; i < numVecs; i++) {
+        int64_t num_to_read
+                = std::min((int64_t)std::max(0, N - i * vtype::numlanes),
+                           (int64_t)vtype::numlanes);
+        typename vtype::opmask_t load_mask = ((0x1ull << num_to_read) - 0x1ull)
+                & ((0x1ull << vtype::numlanes) - 0x1ull);
+        vecs[i] = vtype::mask_loadu(
+                vtype::zmm_max(), load_mask, arr + i * vtype::numlanes);
+    }
+
+// Sort each loaded vector
+#pragma GCC unroll 64
+    for (int i = 0; i < numVecs; i++) {
+        vecs[i] = vtype::sort_vec(vecs[i]);
+    }
+
+    // Run the full merger
+    bitonic_fullmerge_n_vec<vtype, numVecs>(&vecs[0]);
+
+// Unmasked part of the store
+#pragma GCC unroll 64
+    for (int i = 0; i < numVecs / 2; i++) {
+        vtype::storeu(arr + i * vtype::numlanes, vecs[i]);
+    }
+// Masked part of the store
+#pragma GCC unroll 64
+    for (int i = numVecs / 2; i < numVecs; i++) {
+        int64_t num_to_write
+                = std::min((int64_t)std::max(0, N - i * vtype::numlanes),
+                           (int64_t)vtype::numlanes);
+        typename vtype::opmask_t store_mask = ((0x1ull << num_to_write) - 0x1ull)

```

Comment:
`store_mask` is the same as `load_mask`. We can store the `load_mask` in an array and re-use them here rather than recompute.